File: src/codes/Users_Manual/advanced_examples/custom_modulation_0.cs
~using AUTD3Sharp;
~using AUTD3Sharp.Utils;
~using AUTD3Sharp.Gain;
~using static AUTD3Sharp.Units;
using AUTD3Sharp.Derive;

[Modulation]
public partial class Burst
{
    private readonly int _length;

    public Burst(int length)
    {
        _length = length;
        _config = SamplingConfig.Freq(4000u * Hz);
    }

    private EmitIntensity[] Calc(Geometry _)
    {
        var buf = Enumerable.Repeat<EmitIntensity>(EmitIntensity.Min, _length).ToArray();
        buf[0] = EmitIntensity.Max;
        return buf;
    }
}

File: src/codes/Users_Manual/advanced_examples/custom_modulation_0.rs
# extern crate autd3;
# extern crate tokio;
# extern crate autd3_driver;
use autd3::prelude::*;
use autd3_driver::derive::*;

#[derive(Modulation)]
pub struct Burst {
    config: SamplingConfig,
    loop_behavior: LoopBehavior,
}

impl Burst {
    pub fn new() -> Self {
        Self { 
            config: SamplingConfig::Freq(4000 * Hz),
            loop_behavior: LoopBehavior::infinite(),
        }
    }
}

impl Modulation for Burst {
    fn calc(&self, geometry: &Geometry) -> Result<HashMap<usize, Vec<u8>>, AUTDInternalError> {
        Self::transform(geometry, |_| {
            Ok((0..4000)
                .map(|i| if i == 3999 { u8::MAX } else { u8::MIN })
                .collect())
        })
    }
}
# fn main() { 
# }